Table no longer access variable length VO table as bytes string

pllim opened the pull request:
Fix #8716 . This does change a subtle behavior, so it should only go into a major release.

I suspect it would affect packages downstream, e.g., `astroquery.vo_conesearch` or whoever is parsing VO table using `Table.read()` and then decoding bytes to UTF-8 themselves.

Fun fact: Incidentally, I have just tagged all four Toms for review!

TODO:
- [ ] Discuss whether this is the correct approach and the possible side effects (breaking behavior, memory usage, performance).
- [x] Add PR number to change log.

pllim commented on astropy/io/votable/tests/converter_test.py:
```diff
@@ -271,3 +271,8 @@ def test_gemini_v1_2():
     assert table is not None
+
+    tt = table.to_table()
+    assert tt['access_url'][0] == (
```
Currently, without this patch, it returns a `b'http://...'`.

pllim commented on astropy/io/votable/tests/table_test.py:
```diff
@@ -28,7 +28,7 @@ def test_table(tmpdir):
 
     field_types = [
-        ('string_test', {'datatype': 'char', 'arraysize': '*'}),
+        ('string_test', {'datatype': 'unicodeChar', 'arraysize': '*'}),
```
Behavior change in existing test.

taldcroft commented on CHANGES.rst:
```diff
@@ -127,6 +127,9 @@ astropy.stats
 astropy.table
 ^^^^^^^^^^^^^
 
+- Accessing variable length columns from VO table no longer returns bytes
```
This should be in the API changes section.  The more fundamental change is to the table item access behavior of a `Column` of `object` dtype whenever the object is `bytes`.  So in fact the VO table does still return a column of `bytes` objects, but it looks like `str` whenever they get accessed.

pllim replied:
Okay, I solidified the change log entry.

pllim commented on astropy/table/_column_mixins.pyx:
```diff
@@ -57,7 +57,8 @@ cdef inline object base_getitem(object self, object item, item_getter getitem):
 
     try:
-        if value.dtype.char == 'S' and not value.shape:
+        if (isinstance(value, bytes) or
```
I am not sure if this is even the correct solution myself. It was @tomdonaldson who really wanted this. I'm gonna wait for him to chime in first.

pllim commented:
Until I hear back from @tomdonaldson , this is not going forward, so removing milestone for now.

bsipocz commented:
> Until I hear back from @tomdonaldson , this is not going forward, so removing milestone for now.

but surely having a decision by 4.0 is feasible?

taldcroft commented:
I have locally rebased this on master and changed the Cython code update to below, and see tests passing for Table and io.votable:
```
    try:
        if value.dtype.char == 'S' and not value.shape:
            value = value.decode('utf-8', errors='replace')
    except AttributeError:
        if isinstance(value, bytes):
            value = value.decode('utf-8', errors='replace')
    return value
```
With this new flow, I don't see any performance change for `Column` item access.  That is expected because it never gets to the `if isinstance()` code in this case (no attribute error since `value` will be a numpy object with dtype and shape attributes).

I confirmed that there is an API change for table access to an Object column that has a Python `bytes` object inside.  In release it returns `bytes`, with this it will return `str`.  At some level that is more consistent with Table behavior, but in any case that seems like a relatively rare situation.

Takeaway: this is good from my perspective for Table, but would definitely need an API changelog entry.

pllim commented:
Re: https://github.com/astropy/astropy/pull/8745#issuecomment-545152428

@taldcroft , regarding the code change that you posted above -- Do you want me to actually change it to that logic or was it just for your personal testing?

I am not sure what you mean by "it never gets to the `if isinstance()` code in this case" because the reason I added it was to handle the VO use case, so I would think the logic flow would actually get there, since you said you noticed a change in behavior. I'd rather not have the check in `except` block because if I remember correctly, `except` is computationally "expensive" and VO already has its own performance issues without adding more to it.

Hope you can clarify!

p.s. I updated the change log and will push it out shortly.

pllim commented on astropy/table/tests/test_column.py:
```diff
@@ -678,6 +677,25 @@ def test_col_unicode_sandwich_create_from_str(Column):
+    assert not c[0]
+    assert c[1] == 'def'  # Was b'def' before PR 8745
+    assert isinstance(c[1], str)  # Was bytes before PR 8745
+    assert not isinstance(c[1], bytes)  # Was str before PR 8745
```
I noted the behavior change in comments here when compared to `astropy` 3.2.2. Is this behavior change acceptable?

pllim commented on astropy/table/tests/test_column.py:
```diff
@@ -678,6 +677,25 @@ def test_col_unicode_sandwich_create_from_str(Column):
+    assert not isinstance(c[1], bytes)  # Was str before PR 8745
+    assert isinstance(c[:0], table.Column)
+    assert np.all(c[:2] == np.array([None, b'def']))
+    assert not np.all(c[:2] == np.array([None, 'def']))
```
Comparison to `np.array` remains unchanged, unlike above.

taldcroft commented:
> @taldcroft , regarding the code change that you posted above -- Do you want me to actually change it to that logic or was it just for your personal testing?

I did want you to change it to maintain table performance.  Yesterday I thought I found a noticeable hit on column access speed with the `isinstance() in there up front.  But trying to reproduce today I don't see it.  E.g.:
```
col = Column([1, 2, 3])
timeit col[1]
```
or
```
col = Column(np.arange(1e6))
timeit l = [c for c in col]
```

Can you verify that the timing in this branch compared to master is the same?  I'm surprised / confused because it has to do the `isinstance` first and then do the original test as well.  In my testing `isinstance()` takes about 100 ns compared to the column access time of 300 ns in master.  So I would have expected to get about 100 ns more in this branch vs. master, but it came out just the same.  Don't forget the Cython needs to be recompiled in different versions.

If there is no noticeable performance difference then I agree avoiding the `except:` is a good thing.

> I am not sure what you mean by "it never gets to the `if isinstance()` code in this case" 

Sorry, I just meant for normal Table column access of some type like float or int.

tomdonaldson commented:
This change has a side effect on write() that FIELDs with string values now default to an IVOA datatype of "unicodeChar" instead of "char".  This change was noted and accounted for in the unit tests, but it turns out it is not benign.

The output() method of the UnicodeChar converter is different than that for Char.  One impact of this is that when a VOTable with a char FIELD is read, then written, the FIELD is output as a unicodeChar with it original value wrapped by b''.  So when input is:
`...`
`   <FIELD ID="mydata" arraysize="*" datatype="char" name="mydata"/>`
`   <DATA>`
 `   <TABLEDATA><TR><TD>My Data Value</TD></TR></TABLEDATA>`
`...`

the output is:

`...`
`   <FIELD ID="mydata" arraysize="*" datatype="unicodeChar" name="mydata"/>`
`   <DATA>`
 `   <TABLEDATA><TR><TD>b'My Data Value'</TD></TR></TABLEDATA>`
`...`

With this complication, and the lingering concerns about performance and Table impacts beyond VOTable reading, @pllim and I decided we should table this for 4.0 and revisit it for 4.1.

pllim commented:
Trying to see if we can do this without touching the `Table` code and end up deep in 🐰 🕳
